### PR TITLE
#356 Code with windows line endings after formatting still has windows endings

### PR DESCRIPTION
**Pull request by mmatloka**
Related to: https://github.com/olafurpg/scalafmt/issues/365#issuecomment-243346157

### CODE REVIEW

File: core/src/main/scala/org/scalafmt/Scalafmt.scala [hunk 1]
```diff
@@ -39,15 +48,23 @@ object Scalafmt {
...
         }
       }
     } catch {
       // TODO(olafur) add more fine grained errors.
       case NonFatal(e) => FormatResult.Failure(e)
     }
   }
+
+  private[this] def containsWindowsLineEndings(code: String): Boolean =
+    code.contains("\r\n")
```
Comment (olafurpg) [inline core/src/main/scala/org/scalafmt/Scalafmt.scala]: `WindowsLineEnding`?

File: core/src/test/scala/org/scalafmt/LineEndingsTest.scala [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package org.scalafmt
+
+import org.scalafmt.util.DiffAssertions
+import org.scalatest.FunSuite
+
+class LineEndingsTest extends FunSuite with DiffAssertions {
+  val javadocStyle = ScalafmtStyle.default.copy(scalaDocs = false)
```
Comment (olafurpg) [inline core/src/test/scala/org/scalafmt/LineEndingsTest.scala]: Why not default style?

**Comment by mmatloka** [comment]
Corrected, squashed, pushed :)

**Comment by olafurpg** [comment]
One last thought, should we have a flag to disable this new behavior?

**Comment by olafurpg** [comment]
Whoops, I was fiddling with automatic publishing from travis (only possible from master branch) and just remembered that I disabled the tests suite to get faster experiments. Could you rebase on master and re-sync?

**Comment by sjrd** [comment]
> One last thought, should we have a flag to disable this new behavior?

Yes please, so that `scalafmtTest` can assert that all my files have Linux line endings.

**Comment by olafurpg** [comment]
@sjrd how about a flag like `--lineEndings unix/windows/preserve` where preserve will not change any line endings, unix enforces \n and windows enforces \r\n.

This option could be a sealed abstract class `LineEnding` and go into ScalafmtStyle. @mmatloka Here is an example on how to parse a custom type in the Cli: https://github.com/olafurpg/scalafmt/blob/892298eb38eec9590bef732478a7f2d8c0eadec8/cli/src/main/scala/org/scalafmt/cli/Cli.scala#L87

**Comment by mmatloka** [comment]
@olafurpg as `preserve` you mean behaviour implemented in this PR or current master behaviour?

**Comment by olafurpg** [comment]
@mmatloka Hmmm. Excellent question. I think preserve should be the behavior implemented in this PR. The current master branch behavior is buggy.

**Comment by olafurpg** [comment]
@mmatloka For the readme, it should be `@b{preserve}` instead of `@b(preserve)`. With the parentheses it assumes the argument is a variable name.

File: readme/Configuration.scalatex [hunk 3]
```diff
@@ -293,3 +293,14 @@
       function(
         argument1,
         argument2)
+  @sect{--lineEndings}
+    Default: @b(preserve)
```
Comment (olafurpg) [inline readme/Configuration.scalatex]: `@b{preserve}` will do the trick ;)

**Comment by mmatloka** [comment]
I have added the flag. Included in tests and docs, however I was not able to generate readme from scalatex locally, I don't know why yet (``` found   : scalatex.site.Section.Proxy
[error]  required: scalatags.Text.all.Frag``` in not touched areas), so I have not checked how does "look like" the added configuration section. Please feel free to adjust the docs descriptions.

File: readme/Configuration.scalatex (hunk 3 above)
Comment (mmatloka) [inline readme/Configuration.scalatex]: thanks :)

**Comment by olafurpg** [comment]
LGTM. Thanks a lot for this contribution 😄  I expect to release very soon, the wifi is pretty bad at home so I may wait until tomorrow when I go to work.

**Comment by mmatloka** [comment]
Thanks :)

**Comment by olafurpg** [comment]
I am afraid the 0.3.2 release is taking longer than I expected. Is that OK? I have some exciting features in the pipeline that are almost complete and I want to bundle everything in a 0.4.

**Pull request merged**